core/runtime/common/trie_storage_provider_impl.cpp

```diff
@@ -53,96 +52,137 @@ namespace kagome::runtime {
              state_root);
     OUTCOME_TRY(batch, trie_storage_->getPersistentBatchAt(state_root));
+    return setTo(std::move(batch));
+  }
+
+  outcome::result<void> TrieStorageProviderImpl::setTo(
+      std::shared_ptr<storage::trie::TrieBatch> batch) {
+    SL_DEBUG(logger_, "Setting storage provider to new batch");
     child_batches_.clear();
-    persistent_batch_ = std::move(batch);
-    current_batch_ = persistent_batch_;
+    base_batch_ = batch;
+    transaction_stack_.empty();
     return outcome::success();
   }
 
   std::shared_ptr<TrieStorageProviderImpl::Batch>
   TrieStorageProviderImpl::getCurrentBatch() const {
-    return current_batch_;
+    return transaction_stack_.empty() ? base_batch_
+                                      : transaction_stack_.back().main_batch;
   }
 
-  outcome::result<std::shared_ptr<TrieStorageProvider::PersistentBatch>>
-  TrieStorageProviderImpl::getChildBatchAt(const common::Buffer &root_path) {
-    if (!child_batches_.count(root_path)) {
+  outcome::result<std::shared_ptr<storage::trie::TrieBatch>>
+  TrieStorageProviderImpl::getOrCreateChildBatchAt(
+      const common::Buffer &root_path) {
+    if (!transaction_stack_.empty()) {
+      if (auto it = transaction_stack_.back().child_batches.find(root_path);
+          it != transaction_stack_.back().child_batches.end()) {
+        return it->second;
+      }
+    }
+    for (auto transaction_it = transaction_stack_.rbegin();
+         transaction_it != transaction_stack_.rend();
+         transaction_it++) {
+      if (auto it = transaction_it->child_batches.find(root_path);
+          it != transaction_it->child_batches.end()) {
+        return it->second;
+      }
+    }
+    auto child_it = child_batches_.find(root_path);
+    if (child_it == child_batches_.end()) {
       SL_DEBUG(logger_,
-               "Creating new persistent batch for child storage {}",
+               "Creating new base batch for child storage {}",
                root_path.toHex());
-      OUTCOME_TRY(child_root_value, getCurrentBatch()->tryGet(root_path));
-      auto child_root_hash =
-          child_root_value
-              ? common::Hash256::fromSpan(*child_root_value).value()
-              : trie_serializer_->getEmptyRootHash();
-      OUTCOME_TRY(child_batch,
-                  trie_storage_->getPersistentBatchAt(child_root_hash));
-      child_batches_.emplace(root_path, std::move(child_batch));
+      OUTCOME_TRY(child_batch, base_batch_->createChildBatch(root_path));
+      // we check for duplicate batches in the if above
+      BOOST_ASSERT(child_batch.has_value());
+      child_batches_.emplace(root_path, child_batch.value());
+      return child_batch.value();
     }
-    SL_DEBUG(
-        logger_, "Fetching persistent batch for child storage {}", root_path);
-    return child_batches_.at(root_path);
+    SL_DEBUG(logger_, "Fetching batch for child storage {}", root_path);
+    return child_it->second;
```

Comment:
Must fill nested transactions above with topper child before returning.
Bad example of this code:
```
start_tx (tx 1)
child_get (child batch 1 in tx 1)
start_tx (tx 2)
child_put (writes to child batch 1 in tx1 outside of current tx 2)
revert_tx (tx 2)
(tx 2 reverted but child_put result is not reverted)
```

Reply:
I don't think it's true because in order to call child_put we're gonna need to call getMutableChildBatchAt first, which will create the batch for tx 2.

Reply:
Then `getMutableChildBatchAt` must fill missing layers